src/Hamiltonians/FroehlichPolaron.jl

```diff
@@ -0,0 +1,159 @@
+"""
+function FroehlichPolaron(
+    addr::OccupationNumberFS;
+    v=1.0,
+    mass=1.0,
+    omega=1.0,
+    l=1.0,
+    p=0.0,
+    momentum_cutoff=nothing,
+    mode_cutoff=100.0,
+)
+
+The 1D Froehlich polaron Hamiltonian is given by
+
+```math
+H = \\frac{(P̂_f - P)^2}{m} + ωN̂ - ν Σₖ(âₖ^† + â_{-k})
+```
+
+where ``P`` is the total momentum `total_mom`, ``P̂_f = \\frac{M}{L}Σ_k k âₖ^† âₖ`` is the momentum operator for the bosons, ``N̂ = Σ_k âₖ^† âₖ`` is the number operator for the bosons, and ``ν`` is the coupling strength. Here, ``M`` is the number of discrete momentum modes and ```L`` is the momentum scaling.
+
+The `address` must be of type [`OccupationNumberFS`](@ref).
+
+# Arguments
+
+* `addr`: the starting address.
+* `v`: the coupling strength
+* `mass`: the particle mass
+* `omega`: the oscillation frequency of the phonons
+* `l`: the next-neighbor interaction
+* `p`: the total momentum
+* `momentum_cutoff`: the maximum total momentum allowed for a basis element
+* `mode_cutoff`: the maximum population in each momentum mode
+
+
+"""
+struct FroehlichPolaron{
+    T, # eltype
+    M, # number of modes
+    A<:OccupationNumberFS{M}, # address type
+    MC # momentum cutoff indicating type
+} <: AbstractHamiltonian{T}
+    addr::A
+    v::T
+    mass::T
+    omega::T
+    l::T
+    p::T
+    ks::SVector{M,T} # values for k
+    momentum_cutoff::MC
+    mode_cutoff::T
+end
+
+function FroehlichPolaron(
+    addr::OccupationNumberFS;
+    v=1.0,
+    mass=1.0,
+    omega=1.0,
+    l=1.0,
+    p=0.0,
+    momentum_cutoff=nothing,
+    mode_cutoff=10.0,
+)
+    if _exceed_mode_cutoff(mode_cutoff,addr)
+        throw(ArgumentError("Starting address cannot have occupations that exceed mode_cutoff"))
+    end
+
+    M = num_modes(addr) # this is compile-time information
+    if isnothing(momentum_cutoff)
+        v, p, mass, omega, l, mode_cutoff = promote(float(v), float(p), float(mass), float(omega), float(l), float(mode_cutoff))
+    else
+        v, p, mass, omega, l, momentum_cutoff, mode_cutoff = promote(float(v), float(p), float(mass), float(omega), float(l), float(mode_cutoff), float(momentum_cutoff))    
+    end
+
+    step = typeof(v)(2π/M)
+    if isodd(M)
+        start = -π*(1+1/M) + step
+    else
+        start = -π + step
+    end
+    kr = range(start; step = step, length = M)
+    ks = SVector{M}(kr)
+
+    if !isnothing(momentum_cutoff)
+        momentum = (M/l) * dot(ks,onr(addr))
```

Comment:
```suggestion
        momentum = dot(ks,onr(addr))
```
Instead of always having to multiply with `M/l` it might make more sense to set up the momentum grid such that this factor is already in there. That makes it conceptually easier and you even save a few cpu cycles.
This requires changing the setup of the `ks`, though.

Reply:
I've made the changes to complete the (M/l) division directly when the `ks` vector is initialised.